libraries/chez/common/immutable/list.effekt

```diff
@@ -3,124 +3,643 @@ module immutable/list
 import effekt
 import immutable/option
 
+// Used as a type for `Exception` in functions which expect a nonempty list.
+record EmptyList()
+
+// Immutable linked list for finite sequences of elements.
 type List[A] {
   Nil();
   Cons(head: A, tail: List[A])
 }
 
-def map[A, B](l: List[A]) { f: A => B } : List[B] =
-  l match {
-    case Nil() => Nil[B]()
-    case Cons(a, rest) => Cons(f(a), map(rest) { a => f(a) })
+// Create an empty list.
+//
+// O(1)
+def empty[A](): List[A] = Nil()
+
+// Create a list with one element.
+//
+// O(1)
+def singleton[A](x: A): List[A] = Cons(x, Nil())
+
+// Create a list of length `size` where all elements are `default`.
+//
+// O(size)
+def fill[A](size: Int, default: A): List[A] = {
+  build(size) { i => default }
+}
+
+// Create a list from a function `index` of given `size`.
+//
+// O(size)
+def build[A](size: Int) { index: Int => A }: List[A] = {
+  var result = empty()
+  each(0, size) { i =>
+    result = Cons(index(i), result)
   }
+  result.reverse
+}
 
-def foreach[A](l: List[A]) { f: A => Unit } : Unit =
-  l match {
-    case Nil() => ()
-    case Cons(a, rest) =>
-      f(a);
-      foreach(rest) { a => f(a) }
+// Check if list is empty.
+//
+// O(1)
+def isEmpty[A](l: List[A]): Boolean = l match {
+  case Nil() => true
+  case Cons(a, rest) => false
+}
+
+// Check if list is nonempty.
+//
+// O(1)
+def nonEmpty[A](l: List[A]): Boolean = l match {
+  case Nil() => false
+  case Cons(a, rest) => true
+}
+
+// Return the first element of a given list.
+// Throws a `EmptyList` exception if it's empty.
+//
+// O(1)
+def head[A](l: List[A]): A / Exception[EmptyList] = l match {
```

Comment:
Wouldn't the name `unsafeHead` be more appropriate? I guess this is a philosophical question but I would prefer totality by default. The `headOption` could then be named `head`.
Same for `tail`/`unsafeTail`

Reply:
We sometimes use `unsafe` to really mean unsafe (as in "unchecked ffi"). Here, it throws an exception which still can be caught.

I think this is one important point in the design of a stdlib in presence of effects and handlers. Do we want to force users to pattern match or to handle?

Reply:
That's an important point, yes.
In this PR, I didn't change the `head` and `headOption` functions at all in order to side-step the issue :)

However, I added `get` and `unsafeGet` where the former returns an option and the latter panics at runtime (so it's _truly unsafe_, mostly intended for library writers...)
I'm not convinced that we really need the unsafe version though. Having `get` and `getOption` might be more consistent.

I'm very much open to debate here — now that there's a sketch of a library with a few more functions, we get a small view into library design in the language.